presto-main/src/main/java/io/prestosql/execution/buffer/PagesSerde.java

```diff
@@ -56,86 +56,225 @@ public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> co
+                        slice.length(),
+                        decrypted,
+                        0);
+
+                slice = Slices.wrappedBuffer(decrypted, 0, decryptedSize);
+                inUseTempBuffer = decrypted;
+            }
+
+            if (serializedPage.isCompressed()) {
+                checkState(decompressor.isPresent(), "Page is compressed, but decompressor is missing");
+
+                int uncompressedSize = serializedPage.getUncompressedSizeInBytes();
+                byte[] decompressed = context.acquireBuffer(uncompressedSize);
+                checkState(decompressor.get().decompress(
+                        slice.byteArray(),
+                        slice.byteArrayOffset(),
+                        slice.length(),
+                        decompressed,
+                        0,
+                        uncompressedSize) == uncompressedSize);
 
-        if (serializedPage.isEncrypted()) {
-            checkState(spillCipher.isPresent(), "Page is encrypted, but spill cipher is missing");
+                slice = Slices.wrappedBuffer(decompressed, 0, uncompressedSize);
+                if (inUseTempBuffer != null) {
+                    //  Previous buffer is no longer in use
+                    context.releaseBuffer(inUseTempBuffer);
+                }
+                inUseTempBuffer = decompressed;
+            }
+
+            return readRawPage(serializedPage.getPositionCount(), slice.getInput(), blockEncodingSerde);
+        }
+        finally {
+            if (inUseTempBuffer != null) {
+                context.releaseBuffer(inUseTempBuffer);
+            }
+        }
+    }
 
-            byte[] decrypted = new byte[spillCipher.get().decryptedMaxLength(slice.length())];
-            int decryptedSize = spillCipher.get().decrypt(
-                    slice.byteArray(),
-                    slice.byteArrayOffset(),
-                    slice.length(),
-                    decrypted,
-                    0);
+    public static final class PagesSerdeContext
+            implements AutoCloseable
+    {
+        //  Limit retained buffers to 4x the default max page size
+        private static final int MAX_BUFFER_RETAINED_SIZE = DEFAULT_MAX_PAGE_SIZE_IN_BYTES * 4;
+
+        private DynamicSliceOutput sliceOutput;
+        //  Wraps two buffers since encryption + decryption will use at most 2 buffers at once. Buffers are kept in relative order
+        //  based on length so that they can be used for compression or encryption, maximizing reuse opportunities
+        private byte[] largerBuffer;
+        private byte[] smallerBuffer;
+        private boolean closed;
+
+        private void checkNotClosed()
+        {
+            if (closed) {
+                throw new IllegalStateException("PagesSerdeContext is already closed");
+            }
+        }
 
-            slice = Slices.wrappedBuffer(decrypted, 0, decryptedSize);
+        private DynamicSliceOutput acquireSliceOutput(int estimatedSize)
```

Comment:
can slice output be aquired twice?

Reply:
As long as the requested size is smaller than it's current size and it's current size stays smaller than `MAX_BUFFER_RETAINED_SIZE`, then yes- these can be reused in multiple serialization runs which is why even the "no compression, no encryption" case must copy the slice contents before returning.